File: spice3/MNACircuit.py
import numpy as np

from MNAElement import MNAElement, ElementType
from MNASolution import MNASolution
from Unknowns import UnknownCurrent, UnknownVoltage
from Equation import Term, Equation

def get_index_by_equals(array, element) -> int:
    for i in range(len(array)):
        if (type(array[i]) == type(element)) and (array[i] == element):
            return i

    return -1

class MNACircuit:
    def __init__(self, elements: list[MNAElement]):
        self.batteries: list[MNAElement] = []
        self.resistors: list[MNAElement] = []
        self.current_sources: list[MNAElement] = []

        for e in elements:
            if e.type == ElementType.BATTERY:
                self.batteries.append(e)
            elif e.type == ElementType.RESISTOR:
                self.resistors.append(e)
            elif e.type == ElementType.CURRENT_SOURCE:
                self.current_sources.append(e)

        self.elements = self.batteries + self.resistors + self.current_sources

        self.nodes = []

        for e in self.elements:
            self.nodes.append(e.n0)
            self.nodes.append(e.n1)

        self.nodes = list(set(self.nodes))

        self.node_count = len(self.nodes)


    def __str__(self):
        return f"resistors: {map(str, self.resistors)}\nbatteries: {map(str, self.batteries)}\ncurrent sources: {map(str, self.current_sources)}"

    def get_current_count(self):
        free_resistors = 0

        for r in self.resistors:
            if r.value == 0:
                free_resistors += 1

        return len(self.batteries) + free_resistors

    def get_num_vars(self):
        return self.node_count + self.get_current_count()

    def get_current_source_total(self, node_index):
        total = 0.0

        for c in self.current_sources:
            if c.n1 == node_index:
                total -= c.value
            if c.n0 == node_index:
                total += c.value

        return total

    def get_current_terms(self, node: int, side: int, sign: int) -> list[Term]:
        node_terms: list[Term] = []

        for b in self.batteries:
            b_side = b.n0 if side == 0 else b.n1

            if b_side == node:
                node_terms.append(Term(sign, UnknownCurrent(b)))

        for r in self.resistors:
            r_side = r.n0 if side == 0 else r.n1

            if r_side == node and r.value == 0:
                node_terms.append(Term(sign, UnknownCurrent(r)))

            if r_side == node and r.value != 0:
                node_terms.append(Term(-sign/r.value, UnknownVoltage(r.n1)))
                node_terms.append(Term(sign/r.value, UnknownVoltage(r.n0)))

        return node_terms

    def get_ref_node_ids(self) -> list[int]:
        to_visit: list[int] = self.nodes.copy()

        ref_node_ids: list[int] = []

        while len(to_visit) > 0:
            ref_node_id = to_visit[0]
            ref_node_ids.append(ref_node_id)
            connected = self.get_connected_node_ids(ref_node_id)

            for c in connected:
                to_visit.remove(c)

        return ref_node_ids


    def get_connected_node_ids(self, node: int):
        visited = []
        to_visit = [node]

        while len(to_visit) > 0:
            node_to_visit = to_visit.pop(0)

            visited.append(node_to_visit)

            for e in self.elements:
                if e.contains_node_id(node_to_visit):
                    opposite = e.get_opposite_node(node_to_visit)
                    if opposite not in visited:
                        to_visit.append(opposite)

        return list(set(visited))

    def get_equations(self) -> list[Equation]:
        equations = []

        ref_node_ids = self.get_ref_node_ids()

        for r in ref_node_ids:
            equations.append(Equation(0, [
                Term(1, UnknownVoltage(r))
            ]))


        nodes = self.nodes

        for n in nodes:

            if n not in ref_node_ids:
                incoming = self.get_current_terms(n, 1, -1)
                outgoing = self.get_current_terms(n, 0, +1)

                terms = incoming + outgoing

                equations.append(Equation(self.get_current_source_total(n), terms))

        for b in self.batteries:
            equations.append(Equation(b.value, [
                Term(-1, UnknownVoltage(b.n0)),
                Term(1, UnknownVoltage(b.n1))
            ]))

        for r in self.resistors:
            if r.value == 0:
                equations.append(Equation(r.value, [
                    Term(1, UnknownVoltage(r.n0)),
                    Term(-1, UnknownVoltage(r.n1))
                ]))

        return equations


    def get_unknown_currents(self) -> list:
        unknown_currents = []

        for b in self.batteries:
            unknown_currents.append(UnknownCurrent(b))

        for r in self.resistors:
            if r.value == 0:
                unknown_currents.append(UnknownCurrent(r))

        return unknown_currents


    def solve(self):
        equations = self.get_equations()
        unknown_currents = self.get_unknown_currents()
        unknown_voltages = list(map(lambda comp: UnknownVoltage(comp), self.nodes))

        print([str(x) for x in equations])
        print([str(x) for x in unknown_currents])
        print([str(x) for x in unknown_voltages])

        unknowns: list = unknown_currents + unknown_voltages

        print([str(x) for x in unknowns])

        A = np.zeros((len(equations), self.get_num_vars()), dtype=float)
        z = np.zeros((len(equations), 1), dtype=float)

        for row in range(len(equations)):
            print("\n\n")
            print(f"Stamp row {row} with equation {str(equations[row])}")
            print(A)
            print(z)
            print("\n\n")
            equations[row].stamp(row, A, z, lambda comp: get_index_by_equals(unknowns, comp))

        print("\nFinal:")
        print(A, "\n\n", z)

        try:
            x = np.linalg.solve(A, z)
        except np.linalg.LinAlgError:
            print("LinAlgError")
            x = np.zeros((len(equations), 1), dtype=float) # A.n, 1 original

        print("\nOut:\n")
        print(x)

        voltage_map = {}

        print("\nUnknown Voltages:")
        for v in unknown_voltages:
            voltage = x[get_index_by_equals(unknowns, v), 0]
            voltage_map[v.node] = voltage
            print(v.node, "=>", voltage)

        print("\nUnknown Currents:")
        for c in unknown_currents:
            c.element.current_solution = x[get_index_by_equals(unknowns, c), 0]
            print(c.element, c.element.current_solution)

        print("\nVoltage map:")
        print(voltage_map)

        return MNASolution(voltage_map, list(map(lambda comp: comp.element, unknown_currents)))

File: spice2/MNACircuit.cpp
//
// Created by rhys on 18/04/2021.
//

#include "MNACircuit.h"

#include <Eigen/Dense>
#include <iostream>


template<typename T>
T vec_pop_front(std::vector<T>& vec){
    std::reverse(vec.begin(),vec.end()); // first becomes last, reverses the vector
    T out = vec.back();
    vec.pop_back(); // pop last
    std::reverse(vec.begin(),vec.end()); // reverses it again, so the elements are in the same order as before

    return out;
}

MNACircuit::MNACircuit(std::vector<MNAElement *> batteries, std::vector<MNAElement *> resistors, std::vector<MNAElement *> currentSources) {
    this->batteries = batteries;
    this->resistors = resistors;
    this->currentSources = currentSources;

    elements.insert(elements.end(), this->batteries.begin(), this->batteries.end());
    elements.insert(elements.end(), this->resistors.begin(), this->resistors.end());
    elements.insert(elements.end(), this->currentSources.begin(), this->currentSources.end());

    for(auto e : elements){
        nodeSet.insert(std::pair<int, int>{e->n0, e->n0});
        nodeSet.insert(std::pair<int, int>{e->n1, e->n1});
        std::cout << e->n0 << ": " << e->n1 << "\n";
    }
    std::cout << "\n";

    nodeCount = nodeSet.size();

    for(auto n : nodeSet){
        nodes.push_back(n.second);
    }

    for(auto n : nodes){
        std::cout << n << ", ";
    }
    std::cout << "\n";
}

int MNACircuit::getCurrentCount() {
    int freeResistors = 0;
    for(auto r : resistors){
        if(r->value == 0){
            freeResistors++;
        }
    }

    return batteries.size()-freeResistors;
}

int MNACircuit::getNumVars() {
    return nodeCount+getCurrentCount();
}

double MNACircuit::getCurrentSourceTotal(int nodeIndex) {
    double numCurrentSources = 0.0;

    for(auto c : currentSources){
        if(c->n1 == nodeIndex){
            numCurrentSources -= c->value;
        }

        if(c->n0 == nodeIndex){
            numCurrentSources += c->value;
        }
    }

    return numCurrentSources;
}

std::vector<Term *>* MNACircuit::getCurrentTerms(int node, int side, int sign) {
    auto* out = new std::vector<Term*>;

    for(auto b : batteries){
        int bside = side == 0? b->n0 : b->n1;

        if(bside == node) {
            out->push_back(new Term(sign, new UnknownCurrent(b)));
        }
    }

    for(auto r : resistors) {
        int rside = side == 0 ? r->n0 : r->n1;

        if (rside == node && r->value == 0) {
            out->push_back(new Term(sign, new UnknownCurrent(r)));
        }

        if(rside == node && r->value != 0){
            out->push_back(new Term(-sign / r->value, new UnknownVoltage(r->n1 )));
            out->push_back(new Term(sign / r->value, new UnknownVoltage(r->n0 )));
        }
    }

    return out;
}

std::vector<int>* MNACircuit::getReferenceNodeIds() {
    auto* out = new std::vector<int>;

    std::vector<int> toVisit;
    for(auto n : nodeSet){
        toVisit.push_back(n.second);
    }

    while(! toVisit.empty()){
        int refNodeId = toVisit.at(0);

        out->push_back(refNodeId);

        std::vector<int>* connectedNodes = getConnectedNodeIds(refNodeId);

        for(auto c : *connectedNodes){
            toVisit.erase(std::remove_if(toVisit.begin(), toVisit.end(), [c](int i){
                return i == c;
            }));
        }
    }

    return out;
}

std::vector<int> *MNACircuit::getConnectedNodeIds(int node) {
    auto* visited = new std::vector<int>;
    std::vector<int> toVisit = {node};

    while(! toVisit.empty()){
        int nodeToVisit = vec_pop_front<int>(toVisit);

        visited->push_back(nodeToVisit);

        for(auto e : elements){
            if(e->containsNode(nodeToVisit)){
                int oppositeNode = e->getOppositeNode(nodeToVisit);
                if(! std::count(visited->begin(), visited->end(), oppositeNode)){
                    toVisit.push_back(oppositeNode);
                }
            }
        }
    }

    visited->erase(unique( visited->begin(), visited->end() ), visited->end());
    return visited;
}

std::vector<Equation *>* MNACircuit::getEquations() {
    auto* equations = new std::vector<Equation *>;

    std::vector<int>* refNodeIds = getReferenceNodeIds();

    for(auto r : *refNodeIds){
        equations->push_back(new Equation(0, {
            new Term(1, new UnknownVoltage(r))
        }));
    }

    for(auto n : nodes){
        if(! std::count(refNodeIds->begin(), refNodeIds->end(), n)){
            std::vector<Term *>* incoming = getCurrentTerms(n, 1, -1);
            std::vector<Term *>* outgoing = getCurrentTerms(n, 0, +1);

            auto* conserved = new std::vector<Term *>;
            conserved->insert(conserved->end(), incoming->begin(), incoming->end());
            conserved->insert(conserved->end(), outgoing->begin(), outgoing->end());

            equations->push_back(new Equation(getCurrentSourceTotal(n), *conserved));
        }
    }

    for(auto b : batteries){
        equations->push_back(new Equation(b->value, {
            new Term(-1, new UnknownVoltage(b->n0)),
            new Term(1, new UnknownVoltage(b->n1))
        }));
    }

    for(auto r : resistors){
        if(r->value == 0){
            equations->push_back(new Equation(r->value, {
                    new Term(1, new UnknownVoltage(r->n0)),
                    new Term(-1, new UnknownVoltage(r->n1))
            }));
        }
    }

    return equations;
}

std::vector<UnknownCurrent *> *MNACircuit::getUnknownCurrents() {
    auto* out = new std::vector<UnknownCurrent*>;

    for(auto b : batteries){
        out->push_back(new UnknownCurrent(b));
    }

    for(auto r : resistors){
        if(r->value == 0) {
            out->push_back(new UnknownCurrent(r));
        }
    }

    return out;
}

MNASolution *MNACircuit::solve() {
    std::vector<Equation *>* equations = getEquations();
    std::vector<UnknownCurrent*>* unknownCurrents = getUnknownCurrents();
    auto* unknownVoltages = new std::vector<UnknownVoltage*>;

    for(auto v : nodes){
        unknownVoltages->push_back(new UnknownVoltage(v));
    }

    auto* unknowns = new std::vector<Unknown*>(unknownCurrents->begin(), unknownCurrents->end());

    unknowns->insert(unknowns->end(), unknownVoltages->begin(), unknownVoltages->end());


    auto A = Eigen::MatrixXd(equations->size(), getNumVars());
    auto z = Eigen::MatrixXd(equations->size(), 1);

    for(int i=0 ; i<equations->size() ; i++){
        equations->at(i)->stamp(i, &A, &z, [this, unknowns](Unknown* u) {
            return getIndexByEquals(unknowns, u);
        });
    }

    Eigen::MatrixXd x = A.fullPivLu().solve(z); // https://eigen.tuxfamily.org/dox/group__TutorialLinearAlgebra.html

    std::cout << "A:\n" << A << "\n\nz: \n" << z << "\n\nx:" << x << "\n\n\n";

    auto* voltageMap = new std::map<int, double>;

    for(auto v : *unknownVoltages){
        auto rhs = x(getIndexByEquals(unknowns, v));

        voltageMap->insert(std::pair<int, double>(v->node, rhs));
    }

    std::vector<MNAElement*> elems;

    for(auto c : *unknownCurrents){
        c->element->currentSolution = x(getIndexByEquals(unknowns, c), 0);
        elems.push_back(c->element);

    }

    return new MNASolution(*voltageMap, elems);
}

int MNACircuit::getIndexByEquals(std::vector<Unknown*>* array, Unknown* element) {
    for(int i=0 ; i<array->size() ; i++){
        if(array->at(i)->equals(element)){
            return i;
        }
    }

    return -1;
}
